### PR TITLE
Kiali working with no Istiod access

### LINKED ISSUE #5626: Investigate the idea of Kiali working with no istiod connection
Investigate and define how Kiali should work - in a limited way - when there is no connection with Istiod. 

At the moment there is a hard dependency with Istiod, used to retrieve data that Istio has in memory, and also the Istio configuration. 

This is an example (With anonymous auth - special case as it uses the Istio SA to access the resources)

image
image
image

### PR DESCRIPTION
**Pull request by josunect**
Proposal as a first version to have Kiali running with no Istiod. 

Details in: https://docs.google.com/document/d/1XjkodS5Uu14C_1l3-j7iNaWfLufQFycY-cKMlagIxyc

Changes: 

- Not perform some calls if there is no access to istiod. They were making the system really slow (When istiod was not available): 
  - get pod proxy status for workloads
  - Get registry services for services
  - Fetch registry services
- Istio Config: Show a message indicating that Istiod is not available

Some tests to review performance: Frontend tests were showing a degraded performance so a new commit was sent to internally cache the istioAccess value

- [ ] How this could be tested
- [ ] Documentation PR: https://github.com/kiali/kiali.io/pull/613
- [ ] Operator PR: https://github.com/kiali/kiali-operator/pull/595

Fixes #5626

### CODE REVIEW

**Comment by Xunzhuo** [comment]
Great! If you want some reviews, ping me after this is ready! @josunect

**Comment by josunect** [comment]
> Great! If you want some reviews, ping me after this is ready! @josunect

Thanks, @Xunzhuo ! It would be great! I'll let you know.

**Comment by josunect** [comment]
Changes when istiod is not available:

**Istio no available messages**

This remains the same. Should we change it to just a warning instead of error? 
image

**Service mesh version unknown**

image

**Workload status warning because no proxy pod sync**

image

**Istio Config not available**

Show a message instead
image

File: kubernetes/istio.go
Comment (nrfox) [inline kubernetes/istio.go:189]: I think this would be better as a config option for the registry where you can enable/disable the registry altogether if you'd like rather than having it just be a fallback option: https://github.com/kiali/kiali/blob/master/config/config.go#L208.

@hhovsepy may be looking into this but you should be able to get the istio config from the kube api rather than the registry when there's no istiod access. 

What will be the impact on the controlplane namespace card once there is no direct istiod access? It might still be possible to get some details, like controlplane revision, about the controlplane from the injection webhook since this may be present in the cluster even if the controlplane is not.

**Review by nrfox — COMMENTED** [review]
Thanks for picking up this investigation!

File: kubernetes/istio.go
Comment (jshaughn) [reply kubernetes/istio.go:189]: Funny, I was just looking at the same thing.  I also think it makes sense to configure Kiali for this scenario.  By default Kiali would expect to be able to connect to istiod's debug endpoint and proceed normally.  If it can't connect then that would be a FAILURE scenario for the control plane card.  If configured to ignore istiod then it would be a DEGRADED scenario for the control plane card (even if it is accessible we won't connect, this could be for performance reasons, or whatever).

**Review by jshaughn — COMMENTED** [review]
Good progress @josunect , I think I agree with @nrfox about making this a more explicit configuration option for Kiali (see comments).  What do you think?

File: kubernetes/istio.go
Comment (hhovsepy) [reply kubernetes/istio.go:189]: draft solution of retrieving configs when istiod is not available https://github.com/kiali/kiali/pull/5649

**Comment by josunect** [comment]
> Good progress @josunect , I think I agree with @nrfox about making this a more explicit configuration option for Kiali (see comments). What do you think?

Thanks, @jshaughn and @nrfox for the review! I'll approach the changes and submit. 

I've though in do it this way as it can be more robust in the sense that it can work even if Istiod is or is not available, but it is true that it is not possible to know the reason, can be intentional or not - a failure. So it make sense to do it more explicit as part of the configuration.

File: kubernetes/istio.go
Comment (josunect) [reply kubernetes/istio.go:189]: > draft solution of retrieving configs when istiod is not available #5649

Thanks, @hhovsepy !

**Comment by josunect** [comment]
I've also added the following change: 

image

**Review by Xunzhuo — COMMENTED** [review]
I have reviewed the backend codes, did not find any issues. LGTM. Wait for other reviews in front part :)

**Comment by hhovsepy** [comment]
Is it supposed, that the case when 'istiod' is not running/accessible and the config "istio_api_enabled: true" will be handled?  Because it throws "open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory" error message while selecting "istiod" workload.

**Comment by josunect** [comment]
> Is it supposed, that the case when 'istiod' is not running/accessible and the config "istio_api_enabled: true" will be handled? Because it throws "open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory" error message while selecting "istiod" workload.

In that case, it won't be handled - That should be the default behavior. But, I will reproduce it.

File: config/config.go
Comment (nrfox) [inline config/config.go:217]: This option is just related to the registry right? You still get controlplane information even with this disabled? If so, I think it should go under the `Registry` config and then be a bool `Enabled` to be consistent with the other options like `Tracing.Enabled`.

File: business/workloads.go
Comment (nrfox) [inline business/workloads.go:1728]: This might actually port forward or exec directly into the pod to get the proxy status instead of talking with the istiod. Have you tried this without istiod access and it doesn't work?

File: business/istio_validations.go
Comment (nrfox) [inline business/istio_validations.go:106]: We don't want to fetch services or non local mtls configs when the registry is disabled? Seems like we only want to exclude the registry services?

File: config/config.go
Comment (nrfox) [reply config/config.go:217]: Even if istiod access is disabled, I think we can at least infer some controlplane info from different objects in the kube cluster if istiod is deployed there such as the istiod deployment itself or the sidecar injection webhook.

**Comment by jmazzitelli** [comment]
> Many errors in Kiali pod log: `1202 12:56:50.277723 1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:istio-system:kiali-service-account" cannot list resource "pods" in API group "" at the cluster scope

@hhovsepy What is your `deployment.accessible_namespaces` set to? Is it `['**']` or a list of namespaces?

That error message says it wants permission to list pods AT THE CLUSTER SCOPE. Which means it expects (for some reason) a ClusterRole. The Kiali SA only gets bound to a ClusterRole when accessible_namespaces is set to `['**']`. Otherwise, there are just Roles for each namespace listed in accessible_namespaces.

We never before _required_ ClusterRole access to look at pods. So this error message seems strange. We should avoid this requirement if possible.

**Review by hhovsepy — DISMISSED** [review]
Verified now. After correct installation and configuration on OCP cluster.

File: business/workloads.go
Comment (josunect) [reply business/workloads.go:1728]: Thanks, @nrfox, this is a good suggestion. I would try to get the information like this.

File: config/config.go
Comment (josunect) [reply config/config.go:217]: I think it would make sense if we are getting some more controlplane info

File: business/services.go
Comment (jshaughn) [inline business/services.go:486]: Same comment here, maybe just skip the whole go func() block?

File: business/workloads.go
Comment (jshaughn) [reply business/workloads.go:1728]: Although this may work, I'm a little worried about perf when scaling to a lot of pods.  Is that something we should worry about?  Do we already exec into the pod for some other reason?  For me, if the users elect to block istiod access perhaps they can expect to lose this  feature.

**Review by jshaughn — COMMENTED** [review]
A couple of comments.  Also, is there an operator PR for the new config?

File: business/services.go
Comment (josunect) [reply business/services.go]: Updated

File: business/services.go
Comment (josunect) [reply business/services.go]: Updated

File: business/services.go
Comment (josunect) [reply business/services.go]: Updated

File: business/istio_validations.go
Comment (josunect) [reply business/istio_validations.go]: Updated

**Comment by josunect** [comment]
> A couple of comments. Also, is there an operator PR for the new config?

Regarding adding the option into the Registry. This is a comment from the code: 

// RegistryConfig contains configuration for connecting to an external istiod.
// This is used when Kiali should connect to the istiod via a url instead of port forwarding.

https://github.com/kiali/kiali/blob/master/config/config.go#L204 

But with this option, we will be disabling the connection to an external or local Istio. 
Also, some of the options in IstioConfig (The outside configuration) won't take effect if we are disabling the Istiod access, so probably it would make sense to be outside that section?

**Comment by josunect** [comment]
> A couple of comments. Also, is there an operator PR for the new config?

I was trying to figure out the possibility to obtain the proxy status information but: 

- istioctl doesn't seem to be available in the istio-proxy container: 

```
kubectl exec -i -t details-v1-5fdd69c6f9-wrn5b -n bookinfo -c istio-proxy -- /bin/bash

istio-proxy@details-v1-5fdd69c6f9-wrn5b:/$ istioctl  
bash: istioctl: command not found
istio-proxy@details-v1-5fdd69c6f9-wrn5b:/$ find . -name istioctl
find: './etc/ssl/private': Permission denied
find: './var/cache/apt/archives/partial': Permission denied
find: './var/cache/ldconfig': Permission denied
find: './proc/tty/driver': Permission denied
find: './root': Permission denied
istio-proxy@details-v1-5fdd69c6f9-wrn5b:/$ 
```
Because I guess we are not supposed to have access to the remote istiod pod?
But I'm not sure if it is possible to get this information - It looks like this is calculated using some logic in istiod.

**Comment by jshaughn** [comment]
> But with this option, we will be disabling the connection to an external or local Istio.
Also, some of the options in IstioConfig (The outer configuration) won't take effect if we are disabling the Istiod access, so probably it would make sense to be outside that section?

Ah, OK, I didn't have the complete picture.  I am fine with the current setting.

File: business/workloads.go
Comment (nrfox) [reply business/workloads.go:1728]: > For me, if the users elect to block istiod access perhaps they can expect to lose this feature.

@jshaughn this is what I'm wondering. I wasn't sure if the proxy status came from the pod's proxy or from the istiod pod but it looks like it comes from the istiod pod: https://github.com/kiali/kiali/blob/2165abf70c4cfae7d4a08665af6a008329a35ab3/kubernetes/istio.go#L254-L275. In which case this feature _should_ be disabled when istiod access is disabled. Sorry for the noise!

File: config/config.go
Comment (jshaughn) [reply config/config.go:217]: Given the explanation below, I am now fine with this as-is.  Not sure if @nrfox agrees, perhaps he can comment.

File: business/workloads.go
Comment (jshaughn) [reply business/workloads.go:1728]: no worries, it was a good question.

File: config/config.go
Comment (nrfox) [reply config/config.go:217]: I was suggesting `ExternalServices.IstioConfig.Registry.Enabled` since then it'd be consistent with the other external service configs like `ExternalServices.Tracing.Enabled`. Some fields in `IstioConfig` are still relevant even with the registry disabled such as `GatewayAPIClassName` so having a top level `Enabled: false` field would be confusing if other fields in that config are still relevant. If consistency with the other external services is not important since the Istio external service is more unique anyways then I'm good with keeping `IstioApiEnabled`. What do y'all think?

One minor nit if we're keeping `IstioApiEnabled` it should be `IstioAPIEnabled`.

**Review by hhovsepy — APPROVED** [review]
Tested with "istio_api_enabled: false" and istiod pod not running:
Screenshot from 2022-12-15 14-37-17

**Comment by josunect** [comment]
> Tested with "istio_api_enabled: false" and istiod pod not running: Screenshot from 2022-12-15 14-37-17

Thanks for reviewing, @hhovsepy !

File: frontend/src/pages/IstioConfigList/IstioConfigListPage.tsx
Comment (nrfox) [inline frontend/src/pages/IstioConfigList/IstioConfigListPage.tsx:184]: Why is editing istio config disabled when the registry is disabled?

It looks like you can still use the wizards to create/update/delete istio config.

File: frontend/src/components/IstioStatus/IstioStatus.tsx
Comment (nrfox) [inline frontend/src/components/IstioStatus/IstioStatus.tsx:105]: I wonder if this should really be a permanent warning if users have purposefully disabled istiod access by setting the config option?

File: business/workloads.go
Comment (nrfox) [inline business/workloads.go:1728]: Having an empty proxy status will cause workloads to show a warning on their health details:

Screenshot from 2022-12-15 09-47-30

This is more to do with how the frontend is calculating the workload health. I'm not sure we should show the workload as being unhealthy or show a warning when the istiod access has been disabled.

**Comment by hhovsepy** [comment]
For making istio configs edit/update enabled, 'validations.istio.io' webhook needs to be disabled, which is called on k8s side when Kiali calls directly k8s client for creating configs.

**Comment by nrfox** [comment]
> For making istio configs edit/update enabled, 'validations.istio.io' webhook needs to be disabled, which is called on k8s side when Kiali calls directly k8s client for creating configs. Otherwise any k8s api edit/create calls ends "Internal error occurred: failed calling webhook "validation.istio.io": failed to call webhook:"

What webhook is that?

**Comment by josunect** [comment]
> > For making istio configs edit/update enabled, 'validations.istio.io' webhook needs to be disabled, which is called on k8s side when Kiali calls directly k8s client for creating configs. Otherwise any k8s api edit/create calls ends "Internal error occurred: failed calling webhook "validation.istio.io": failed to call webhook:"
> 
> What webhook is that?

I think is ValidatingWebhookConfiguration: 

```
kubectl get ValidatingWebhookConfiguration
NAME                           WEBHOOKS   AGE
istio-validator-istio-system   1          3d2h
istiod-default-validator       1          3d2h
```

But, i should say this issue did not happen in my local environment when creating an object. I will try to reproduce

File: business/workloads.go
Comment (nrfox) [reply business/workloads.go:1728]: @josunect I still don't think we should show a warning label here for the empty proxy status when istiod access is disabled. As it is, pods will always be in a "warning" state but if you've purposefully disabled registry access then imo the frontend shouldn't try to use proxy status for the health calculation if it has no way of ever getting it.

File: frontend/src/pages/Overview/OverviewPage.tsx
Comment (nrfox) [inline frontend/src/pages/Overview/OverviewPage.tsx:900]: I think this label looks great!

File: frontend/src/pages/Overview/OverviewPage.tsx
Comment (nrfox) [inline frontend/src/pages/Overview/OverviewPage.tsx:943]: Minor: You don't need to resolve this as part of this PR but there's now a bunch of whitespace on the controlplane card without the metrics:

Screenshot from 2022-12-19 21-02-48

I'm not sure what to fill that space with.

**Comment by hhovsepy** [comment]
So when istiod is not available, this means it is not installed, thus "ValidatingWebhookConfiguration istiod-default-validator" is also not created.
So the solution to bring edit back for Istion configs can be merged: https://github.com/josunect/kiali/pull/6

**Comment by josunect** [comment]
> So when istiod is not available, this means it is not installed, thus "ValidatingWebhookConfiguration istiod-default-validator" is also not created. So the solution to bring edit back for Istion configs can be merged: [josunect#6](https://github.com/josunect/kiali/pull/6)

Thanks, @hhovsepy ! It has been merged

**Review by nrfox — COMMENTED** [review]
The only remaining issue I can see is that validations appear to be somewhat degraded with the istio api disabled. As an example, edit a destination rule's subset to match some non-existent label and you'll see a validation error with registry enabled but see no error with the registry disabled. Perhaps some validations need info from the registry? 

@hhovsepy not sure if validations should be degraded or disabled entirely with the registry disabled? If they can still provide some value without the registry imo it would be worth keeping but perhaps there should be a note saying that validations are degraded without registry access? What is your opinion @hhovsepy @josunect?

**Comment by hhovsepy** [comment]
@nrfox the idea of making configs read only was, that Kiali validations will be degraded, because they require registry services. Also, when istiod does not exist, the Istio validations are not performed, so the user can break the configs even without noticing.
There can be a message notifying to user in Istio config YAML edit page that validations are degraded because istiod api is disabled, edit the config on your risk.

**Comment by nrfox** [comment]
> @nrfox the idea of making configs read only was, that Kiali validations will be degraded, because they require registry services. Also, when istiod does not exist, the Istio validations are not performed, so the user can break the configs even without noticing.

Ah I'm sorry then for I didn't realize that the read only mode for istio config was to workaround the validations being degraded when I suggested adding back editing the configs.

> There can be a message notifying to user in Istio config YAML edit page that validations are degraded because istiod api is disabled, edit the config on your risk.

This sounds fine to me. I'm also good with making the istio configs read only again for now if that's what you'd prefer for the validations. Perhaps some of the validations that rely on the registry can still be performed in the future without it. For example, it seems like Kiali should be able to validate Destination Rules matching workloads without needing the registry but this doesn't work today. That's a separate issue though and will require its own investigation.

**Comment by josunect** [comment]
I think we can put a message to let the user know, and still have the possibility to be able to create/edit/remove Istio Configs. Because, this still will be possible to do it manually, right?

**Comment by hhovsepy** [comment]
Yeah agree to put a message on editable configs, and have a separate task for validations: https://github.com/kiali/kiali/issues/5676

**Comment by josunect** [comment]
A message like this? 

image

File: frontend/src/pages/Overview/OverviewPage.tsx
Comment (nrfox) [inline frontend/src/pages/Overview/OverviewPage.tsx:867]: Minor: it looks like the controlplane card's height doesn't match the other cards now with the istio api disabled:
Screenshot from 2023-01-09 12-41-30

**Review by nrfox — COMMENTED** [review]
A few minor comments otherwise LGTM. Looks like you'll need to rebase with the main branch before this can be merged.

**Comment by jshaughn** [comment]
@josunect Please rebase against and then let's proceed with a final review!

**Comment by josunect** [comment]
It will need to test further after the merge. I see some errors like: 

```
2023-01-10T15:01:54Z WRN Error getting proxy status from istiod. Proxy status may be stale. Err: unable to proxy Istiod pods. Make sure your Kubernetes API server has access to the Istio control plane through 8080 port

2023-01-10T15:03:54Z ERR Failed to call Istiod endpoint /debug/syncz error: unable to proxy Istiod pods. Make sure your Kubernetes API server has access to the Istio control plane through 8080 port
```

File: frontend/src/components/IstioWizards/ServiceWizard.tsx
Comment (nrfox) [inline frontend/src/components/IstioWizards/ServiceWizard.tsx:740]:
```suggestion
              <KialiIcon.Warning /> <b>Istio API is disabled.</b> Be careful when editing the configuration as the Istio config validations are disabled when the Istio API is disabled.
```

Comment (nrfox) [reply frontend/src/components/IstioWizards/ServiceWizard.tsx:740]: Missed this in my last review but not a blocker to make it consistent with the other messages.

**Review by nrfox — APPROVED** [review]

File: kubernetes/cache/proxy_status.go
Comment (nrfox) [inline kubernetes/cache/proxy_status.go:22]: Can you move this if statement up to where the function is called in `NewKialiCache` https://github.com/kiali/kiali/blob/master/kubernetes/cache/cache.go#L210-L211? Calling a function that actually does nothing if you've enabled a config option is an odd side effect and could be confusing later on.

**Review by nrfox — APPROVED** [review]
LGTM

**Pull request merged**
